### PR TITLE
REL-2232 fix nifs partitioner

### PR DESCRIPTION
**Pull request by tpolecat**
NIFS targets were not being partitioned correctly, so I fixed that and added some support for end-to-end testing of skeleton generation. See the included [`REL_2232_Test.scala`](https://github.com/tpolecat/ocs/blob/nifs-partitioner/bundle/edu.gemini.phase2.skeleton.servlet/src/test/scala/edu/gemini/phase2/skeleton/factory/REL_2232_Test.scala) for an example. It is likely that the template expansion testing support will be further generalized when I update the Graces tests (after this is merged).

### CODE REVIEW

**Comment by cquiroz** [comment]
Looks good :+1:

**Comment by tpolecat** [comment]
Lots of new stuff, needs new review. Sorry.

File: bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/immutable/Observation.scala [hunk 1]
```diff
@@ -11,6 +11,7 @@ object Observation {
   val condition:Lens[Observation, Option[Condition]] = Lens.lensu((a, b) => a.copy(condition = b), _.condition)
   val target:Lens[Observation, Option[Target]] = Lens.lensu((a, b) => a.copy(target = b), _.target)
   val meta:Lens[Observation, Option[ObservationMeta]] = Lens.lensu((a, b) => a.copy(meta = b), _.meta)
+  val time:Lens[Observation, Option[TimeAmount]] = Lens.lensu((a, b) => a.copy(time = b), _.time)
```
Comment (cquiroz) [inline bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/immutable/Observation.scala]: This is unrelated to this PR. What's the state of the art on generating (e.g. via Macros) lenses, I remember some discussions about that

File: bundle/edu.gemini.phase2.skeleton.servlet/skeleton.sh [hunk 2]
```diff
@@ -0,0 +1,106 @@
+#!/bin/bash
```
Comment (cquiroz) [inline bundle/edu.gemini.phase2.skeleton.servlet/skeleton.sh]: I thought `skeleton.sh` was present somewhere on the codebase (It may have been on the svn repo), but in any case great to have it here

File: bundle/edu.gemini.phase2.skeleton.servlet/src/main/scala/edu/gemini/phase2/skeleton/servlet/SkeletonServlet.scala [hunk 3]
```diff
@@ -104,6 +104,10 @@ final class SkeletonServlet(odb: IDBDatabaseService, templateFactory: TemplateFa
         a => a.asInstanceOf[FileItem]
       })
 
+  /** If `test` is set then the PDF file is optional and ignored if present. */
+  private def test(items: List[FileItem]): Boolean =
+    items.exists(_.getFieldName == "test")
```
Comment (cquiroz) [inline bundle/edu.gemini.phase2.skeleton.servlet/src/main/scala/edu/gemini/phase2/skeleton/servlet/SkeletonServlet.scala]: Where does this field come from, is it int the proposal?

File: bundle/edu.gemini.phase2.skeleton.servlet/src/test/scala/edu/gemini/phase2/skeleton/factory/TemplateSpec.scala [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+/** 
+ * Mixin for a specs test for template expansion. This provides code to expand a Phase 1 Proposal
+ * into a skeleton, then test that the expansion was correct.
+ * @param xmlName name of the template to test: "NIFS_BP.xml" for example
+ */
+abstract class TemplateSpec(xmlName: String) { this: Specification =>
+
+  // Required when constructing phase 1 proposals
+  if (System.getProperty("edu.gemini.model.p1.schemaVersion") == null)
```
Comment (cquiroz) [inline bundle/edu.gemini.phase2.skeleton.servlet/src/test/scala/edu/gemini/phase2/skeleton/factory/TemplateSpec.scala]: Hmm, I thought this was removed, and we were reading it from a class, but perhaps I'm talking about the semester

File: bundle/edu.gemini.phase2.skeleton.servlet/src/test/scala/edu/gemini/phase2/skeleton/factory/TemplateSpec.scala [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+  /** Construct an empty phase-1 sidereal target with the given magnitudes. */
+  def p1Target(ms: List[Magnitude]): SiderealTarget =
+    SiderealTarget.empty.copy(name = "test", magnitudes = ms)
+
+  /** Construct a phase-1 observation with the given blueprint and target. */
+  def p1Obs(bp: BlueprintBase, st: Target): Observation =
+    Observation.empty execState {
+      for {
+        _ <- blueprint := Some(bp)
```
Comment (cquiroz) [inline bundle/edu.gemini.phase2.skeleton.servlet/src/test/scala/edu/gemini/phase2/skeleton/factory/TemplateSpec.scala]: This is new to me, what`s the `:=`?

**Comment by cquiroz** [comment]
Great to have the template testing :+1: 
If I understand it correctly you can skip the pdf processing on the tests with these modifications, right?

File: bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/immutable/Observation.scala (hunk 1 above)
Comment (tpolecat) [inline bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/immutable/Observation.scala]: State of the art is a library by Julien Truffaut called [Monocle](https://github.com/julien-truffaut/Monocle). I would like to integrate it once we have upgraded to scalaz 7.1 (or 7.2, depending on when it happens).

File: bundle/edu.gemini.phase2.skeleton.servlet/src/main/scala/edu/gemini/phase2/skeleton/servlet/SkeletonServlet.scala (hunk 3 above)
Comment (tpolecat) [inline bundle/edu.gemini.phase2.skeleton.servlet/src/main/scala/edu/gemini/phase2/skeleton/servlet/SkeletonServlet.scala]: I added it to the servlet so you can skip the PDF and attachment processing. Thanks for pointing this out; I had forgotten about it and need to write a test.

File: bundle/edu.gemini.phase2.skeleton.servlet/src/test/scala/edu/gemini/phase2/skeleton/factory/TemplateSpec.scala (hunk 4 above)
Comment (tpolecat) [inline bundle/edu.gemini.phase2.skeleton.servlet/src/test/scala/edu/gemini/phase2/skeleton/factory/TemplateSpec.scala]: There may be a better way to do this, but I was unable to construct a `Proposal` without it.

File: bundle/edu.gemini.phase2.skeleton.servlet/src/test/scala/edu/gemini/phase2/skeleton/factory/TemplateSpec.scala (hunk 5 above)
Comment (tpolecat) [inline bundle/edu.gemini.phase2.skeleton.servlet/src/test/scala/edu/gemini/phase2/skeleton/factory/TemplateSpec.scala]: `Lens[A,B]` can also be thought of as a pair of operations `get: State[A,B]` and `set: B => State[A, Unit]` and the second is what the `:=` does ... it just lets you do a bunch of lens operations as a program in `State`, which is sometimes easier to follow. Ed Kmett's [lecture](https://www.youtube.com/watch?v=efv0SQNde5Q) here explains this style of lens if you're interested.

**Comment by tpolecat** [comment]
@cquiroz yes, the testing stuff at this point doesn't do any attachment processing; it's all in-memory and has no file IO at all.

**Comment by tpolecat** [comment]
I removed the change to the skeleton servlet because it's not relevant here, and renamed the `testing` parameter to `preserveLibraryIds` which is more descriptive.
